### PR TITLE
libnetwork/drivers/overlay: parse VNI list option in O(1) memory

### PR DESCRIPTION
**Pull request by corhere**
**- What I did**
Follow up of https://github.com/moby/moby/pull/45882#discussion_r1253594377

Modified the VNI list option parsing in the overlay driver to be more memory efficient.

**- How I did it**
Extracted the parsing logic to a common utility function, wrote tests, and modified the implementation to use `strings.Cut` instead of `strings.Split`.

**- How to verify it**
New unit tests assert that the new implementation behaves the same as the old while not allocating memory.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: libnetwork/drivers/overlay/overlayutils/utils_test.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+			name:    "InvalidVNI2",
+			slice:   []uint32{4, 5, 6},
+			csv:     "abc,1,2,3",
+			want:    []uint32{4, 5, 6},
+			wantErr: "invalid vxlan id value \"abc\" passed",
+		},
+	}
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := AppendVNIList(tt.slice, tt.csv)
```
Comment (thaJeztah) [inline libnetwork/drivers/overlay/overlayutils/utils_test.go:58]: While we're writing this with a table test, should we run with `t.Parallel()`?

File: libnetwork/drivers/overlay/ovmanager/ovmanager.go [hunk 2]
```diff
@@ -89,14 +89,10 @@ func (d *driver) NetworkAllocate(id string, option map[string]string, ipV4Data,
...
-			valStrList := strings.Split(val, ",")
-			for _, idStr := range valStrList {
-				vni, err := strconv.Atoi(idStr)
-				if err != nil {
-					return nil, fmt.Errorf("invalid vxlan id value %q passed", idStr)
-				}
-
-				vxlanIDList = append(vxlanIDList, uint32(vni))
+			var err error
+			vxlanIDList, err = overlayutils.AppendVNIList(vxlanIDList, val)
```
Comment (cpuguy83) [inline libnetwork/drivers/overlay/ovmanager/ovmanager.go:93]: I was thinking more along the lines of only parsing each value as needed down at 109.

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
(oh, race condition with Brian's comment; I'll wait for that to proceed)

File: libnetwork/drivers/overlay/overlayutils/utils_test.go (hunk 1 above)
Comment (corhere) [reply libnetwork/drivers/overlay/overlayutils/utils_test.go:58]:
```
ok  	github.com/docker/docker/libnetwork/drivers/overlay/overlayutils	0.006s
```
The tests take so little time to run that the synchronization and scheduling overhead of running them in parallel would at best make it a wash or at worst make things slower.

File: libnetwork/drivers/overlay/ovmanager/ovmanager.go (hunk 2 above)
Comment (corhere) [reply libnetwork/drivers/overlay/ovmanager/ovmanager.go:93]: This PR is already a premature optimization for a code path which will see one, maybe two elements for any practical real-world usage. This is as far as I'm willing to go.

Comment (cpuguy83) [reply libnetwork/drivers/overlay/ovmanager/ovmanager.go:92]: For me it is not so much optimization as it is safety.
But I see your point. I don't really see this particular change as an improvement for performance or safety.

**Review by neersighted — APPROVED** [review]
I think this is a fine middle ground for now; I agree that this code realistically is not nearly as open-ended as the implementation makes it look, and we're better off just obsoleting all this (questionable) logic in the long run.

**Pull request merged**